utils/frame/try-runtime/cli/src/lib.rs

```diff
@@ -53,169 +72,283 @@ pub struct TryRuntimeCmd {
+				modules: modules.to_owned().unwrap_or_default(),
+				at: block_at.as_ref()
+					.map(|b| b.parse().map_err(|e| format!("Could not parse hash: {:?}", e))).transpose()?,
+				..Default::default()
+			})),
+		};
+
+		// inject the code into this ext.
+		builder.inject(&[(code_key, code)]).build().await?
 	};
 
-	if let Some(pos) = block_number.chars().position(|c| !c.is_ascii_hexdigit()) {
-		Err(format!(
-			"Expected block hash, found illegal hex character at position: {}",
-			2 + pos,
-		))
-	} else {
-		Ok(block_number.into())
-	}
+	let encoded_result = StateMachine::<_, _, NumberFor<B>, _>::new(
+		&ext.backend,
+		None,
+		&mut changes,
+		&executor,
+		"TryRuntime_on_runtime_upgrade",
+		&[],
+		ext.extensions,
+		&sp_state_machine::backend::BackendRuntimeCode::new(&ext.backend)
+			.runtime_code()?,
+		sp_core::testing::TaskExecutor::new(),
+	)
+	.execute(execution.into())
+	.map_err(|e| format!("failed to execute 'TryRuntime_on_runtime_upgrade' due to {:?}", e))?;
+
+	let (weight, total_weight) = <(u64, u64) as Decode>::decode(&mut &*encoded_result)
+		.map_err(|e| format!("failed to decode output due to {:?}", e))?;
+	log::info!(
+		"TryRuntime_on_runtime_upgrade executed without errors. Consumed weight = {}, total weight = {} ({})",
+		weight,
+		total_weight,
+		weight as f64 / total_weight as f64
+	);
+
+	Ok(())
 }
 
-fn parse_url(s: &str) -> Result<String, &'static str> {
-	if s.starts_with("ws://") || s.starts_with("wss://") {
-		// could use Url crate as well, but lets keep it simple for now.
-		Ok(s.to_string())
-	} else {
-		Err("not a valid WS(S) url: must start with 'ws://' or 'wss://'")
-	}
+async fn offchain_worker<B, ExecDispatch>(
+	shared: SharedParams,
+	command: OffchainWorkerCmd,
+	config: Configuration,
+)-> sc_cli::Result<()>
+where
+	B: BlockT,
+	B::Hash: FromStr,
+	B::Header: serde::de::DeserializeOwned,
+	<B::Hash as FromStr>::Err: Debug,
+	NumberFor<B>: FromStr,
+	<NumberFor<B> as FromStr>::Err: Debug,
+	ExecDispatch: NativeExecutionDispatch + 'static,
+{
+	let wasm_method = shared.wasm_method;
+	let execution = shared.execution;
+
+	let mut changes = Default::default();
+	let max_runtime_instances = config.max_runtime_instances;
+	let heap_pages = config.default_heap_pages;
+	let executor = NativeExecutor::<ExecDispatch>::new(
+		wasm_method.into(),
+		heap_pages,
+		max_runtime_instances,
+	);
+
+	let (online_config, url) = match command.state {
+			State::Live {
+				url,
+				snapshot_path,
+				block_at,
+				modules
+			} => {
+				let online_config = OnlineConfig {
+				transport: url.to_owned().into(),
+				state_snapshot: snapshot_path.as_ref().map(SnapshotConfig::new),
+				modules: modules.to_owned().unwrap_or_default(),
+				at: block_at.as_ref()
+					.map(|b| b.parse().map_err(|e| format!("Could not parse hash: {:?}", e))).transpose()?,
+				..Default::default()
+				};
+
+				(online_config, url)
+			},
+			_ => {
+				panic!("Only live state is supported");
```

Comment:
btw I am fine with this for now.

Reply:
> Add State::Offline support (Check to see if block header can be extracted from state)

it would have been sensible if this was given, but it is not, and it will add some confusion. For now I think only live is fine, and should work more toward having archive nodes ready to play with (which we currently _almost_ have)

Reply:
I actually adding it because my internet rn is extremely slow. It has the hack right now that it will query `ws://localhost:9944` for the header. I am think I can do follow PR with a little more elegant solution